pam/gdm/protocol_test.go

```diff
@@ -0,0 +1,687 @@
+			gdmData, err := NewDataFromJSON(bytes)
+			require.NoError(t, err)
+			requireEqualData(t, tc.gdmData, gdmData)
+		})
+	}
+}
+
+func TestGdmStructsUnMarshal(t *testing.T) {
+	t.Parallel()
+	tests := map[string]struct {
+		JSON string
+
+		wantData   *Data
+		wantErrMsg string
+	}{
+		"hello packet": {
+			JSON: `{"type":"hello"}`,
+
+			wantData: &Data{Type: DataType_hello},
+		},
+
+		"Hello packet with data": {
+			JSON: `{"type":"hello","hello":{"version":55}}`,
+
+			wantData: &Data{Type: DataType_hello, Hello: &HelloData{Version: 55}},
+		},
+
+		"Event packet": {
+			JSON: `{"type":"event","event":{"type":"brokerSelected","brokerSelected":{}}}`,
+
+			wantData: &Data{
+				Type: DataType_event,
+				Event: &EventData{
+					Type: EventType_brokerSelected,
+					Data: &EventData_BrokerSelected{},
+				},
+			},
+		},
+
+		"Event ack packet": {
+			JSON: `{"type":"eventAck"}`,
+
+			wantData: &Data{Type: DataType_eventAck},
+		},
+
+		"Request packet": {
+			JSON: `{"type":"request","request":{"type":"uiLayoutCapabilities","uiLayoutCapabilities":{}}}`,
+
+			wantData: &Data{
+				Type: DataType_request,
+				Request: &RequestData{
+					Type: RequestType_uiLayoutCapabilities,
+					Data: &RequestData_UiLayoutCapabilities{},
+				},
+			},
+		},
+
+		"Request packet with missing data": {
+			JSON: `{"type":"request","request":{"type":"updateBrokersList"}}`,
+
+			wantData: &Data{
+				Type: DataType_request,
+				Request: &RequestData{
+					Type: RequestType_updateBrokersList,
+				},
+			},
+		},
+
+		"Response packet": {
+			JSON: `{"type":"response","response":{"type":"uiLayoutCapabilities","uiLayoutCapabilities":{}}}`,
+
+			wantData: &Data{
+				Type: DataType_response,
+				Response: &ResponseData{
+					Type: RequestType_uiLayoutCapabilities,
+					Data: &ResponseData_UiLayoutCapabilities{},
+				},
+			},
+		},
+
+		"Response packet with ack data": {
+			JSON: `{"type":"response","response":{"type":"changeStage","ack":{}}}`,
+
+			wantData: &Data{
+				Type: DataType_response,
+				Response: &ResponseData{
+					Type: RequestType_changeStage,
+					Data: &ResponseData_Ack{},
+				},
+			},
+		},
+
+		"Poll packet": {
+			JSON: `{"type":"poll"}`,
+
+			wantData: &Data{Type: DataType_poll},
+		},
+
+		"PollResponse packet": {
+			JSON: `{"type":"pollResponse","pollResponse":` +
+				`[{"type":"brokerSelected","brokerSelected":{"brokerId":"a broker"}}]}`,
+
+			wantData: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{
+						Type: EventType_brokerSelected,
+						Data: &EventData_BrokerSelected{
+							BrokerSelected: &Events_BrokerSelected{BrokerId: "a broker"},
+						},
+					},
+				},
+			},
+		},
+
+		"PollResponse packet with missing data": {
+			JSON: `{"type":"pollResponse"}`,
+
+			wantData: &Data{
+				Type:         DataType_pollResponse,
+				PollResponse: nil,
+			},
+		},
+
+		// Error cases
+		"Error empty packet ": {
+			wantErrMsg: "syntax error",
+		},
+
```

Comment:
This empty line is fine. It's separating the test cases scope, so it should be kept.